https://github.com/rust-lang/rust-clippy/pull/14599 has a few interesting testcases

The cases which are incorrectly linting and #14599 does not lint are all unrelated to this change and already fail on main.

<details><summary>testcases that I am ignoring becauese does not compile under current rust (not sure about polonius, have not tried)</summary>
<p>

I am not, because that would not compile with `error[E0382]: use of moved value: x`

```rust
let mut x = black_box(String::new());
for _ in 0..10 {
    let _y = x.clone(); //~ redundant_clone
    black_box(&mut x);
}
```

I am not, because that would not compile with `error[E0382]: borrow of moved value: y`

```rust
let mut x = black_box(String::new());
for _ in 0..10 {
    let y = black_box(String::new());
    x = y.clone(); //~ redundant_clone
    black_box(&y);
}
```

```rust
let mut x = black_box(String::new());
for _ in 0..10 {
    let y = black_box(String::new());
    x = y.clone(); //~ redundant_clone
    black_box(&y);
}
black_box(&x);
```

</p>
</details>

🤔 The loop-cases which I AM missing (=not linting) are all closely related to aliasing

```rust
let mut x = black_box(String::new());
for _ in 0..10 {
    let y = x; //~ redundant_clone
    black_box(y);
    x = black_box(String::new());
}
```

```rust
let mut x = black_box(String::new());
for _ in 0..10 {
    let y = x.clone(); //~ redundant_clone
    black_box(y);
    x = black_box(String::new());
}
```

```rust
let mut x = black_box(String::new());
let mut y = black_box(String::new());
for _ in 0..10 {
    y = x.clone(); //~ redundant_clone
    x = black_box(String::new());
}
black_box(&x);
```

```rust
let mut x = black_box(String::new());
let mut y = x.clone(); //~ redundant_clone
for _ in 0..10 {
    black_box(y);
    x = black_box(String::new());
    y = x.clone(); //~ redundant_clone
}
```

--- a/clippy_utils/src/mir/mod.rs
+++ b/clippy_utils/src/mir/mod.rs
@@ -1,13 +1,13 @@
-use std::iter;
-use std::ops::Range;
+use std::{iter, mem};
 
 use rustc_data_structures::either::Either;
-use rustc_data_structures::fx::FxHashSet;
 use rustc_hir::{Expr, HirId};
+use rustc_index::IndexVec;
 use rustc_index::bit_set::DenseBitSet;
 use rustc_middle::mir::visit::{MutatingUseContext, NonMutatingUseContext, PlaceContext, Visitor};
 use rustc_middle::mir::{
-    BasicBlock, Body, InlineAsmOperand, Local, Location, Place, START_BLOCK, Statement, StatementKind, TerminatorKind,
+    BasicBlock, BasicBlockData, Body, InlineAsmOperand, Local, Location, Place, START_BLOCK, Statement, StatementKind,
+    TerminatorKind,
 };
 use rustc_middle::ty::TyCtxt;
 
@@ -31,7 +31,7 @@ pub fn visit_local_usage<const N: usize>(
     mir: &Body<'_>,
     location: Location,
 ) -> Option<[LocalUsage; N]> {
-    let blocks = reachable_while_storage_live(&locals, mir, location)?;
+    let live_on_entry = reachable_while_storage_live(&locals, mir, location)?;
 
     let mut v = V {
         locals: &locals,
@@ -45,7 +45,7 @@ pub fn visit_local_usage<const N: usize>(
     };
 
     for &tbb in mir.basic_blocks.reverse_postorder() {
-        if blocks.contains(tbb) {
+        if live_on_entry[tbb] != 0 {
             v.visit_basic_block_data(tbb, &mir.basic_blocks[tbb]);
         }
     }
@@ -77,54 +77,54 @@ fn reachable_while_storage_live<const N: usize>(
     mir: &Body<'_>,
     location: Location,
 ) -> Option<DenseBitSet<BasicBlock>> {
-    let remove_killed = |bb: BasicBlock, range: Range<usize>, live: &mut Live| {
-        for slot in mir.basic_blocks[bb].statements[range]
-            .iter()
-            .filter_map(killed_local)
-            .filter_map(|killed| locals.iter().position(|&local| local == killed))
-        {
-            live.remove(slot);
-        }
-    };
-    let successors = |bb: BasicBlock, live: &Live| {
-        mir.basic_blocks[bb]
-            .terminator()
-            .successors()
-            .map(|succ| (succ, live.clone()))
-            .collect::<Vec<_>>()
-    };
-
-    let mut blocks = DenseBitSet::new_empty(mir.basic_blocks.len());
-    blocks.insert(location.block);
-
-    let mut seen = FxHashSet::default();
-    let mut worklist = Vec::new();
-    // `location.block` is the only block entered part-way through, so it is walked separately.
-    // Reaching it again in the loop below means a cycle closed around `location`.
-    let stmts = mir.basic_blocks[location.block].statements.len();
-    let after_location = (location.statement_index + 1).min(stmts);
-    let mut live = Live::new_filled(N);
-    remove_killed(location.block, after_location..stmts, &mut live);
-    if !live.is_empty() {
-        worklist.extend(successors(location.block, &live));
+    fn join(base: &mut u8, other: u8) -> bool {
+        let new = *base | other;
+        mem::replace(base, new) != new
+    }
+    const ENQUEUED_FLAG: u8 = 0b1000_0000;
+    fn enqueue(state: &mut u8) -> bool {
+        let new = *state | ENQUEUED_FLAG;
+        mem::replace(state, new) != new
+    }
+    fn dequeue(state: &mut u8) {
+        *state &= !ENQUEUED_FLAG;
     }
 
-    while let Some((bb, mut live)) = worklist.pop() {
-        if bb == location.block {
-            return None;
-        }
-        if !seen.insert((bb, live.clone())) {
-            continue;
-        }
-        blocks.insert(bb);
+    let mut states = IndexVec::from_raw(vec![0; mir.basic_blocks.len()]);
+    let mut queue = Vec::new();
+    let init = !(0u8 << N);
+    states[location.block] = init;
 
-        remove_killed(bb, 0..mir.basic_blocks[bb].statements.len(), &mut live);
-        if !live.is_empty() {
-            worklist.extend(successors(bb, &live));
+    let bb_data = &mir.basic_blocks[location.block];
+    let result = apply_deaths(bb_data, location.statement_index + 1, init);
+    if result != 0 {
+        for succ in mir.basic_blocks[location.block].terminator().successors() {
+            if succ == location.block {
+                return None;
+            }
+            if join(&mut states[succ], result) && enqueue(&mut states[succ]) {
+                queue.push(succ);
+            }
         }
     }
 
-    Some(blocks)
+    while let Some(bb) = queue.pop() {
+        dequeue(&mut states[bb]);
+        let bb_data = &mir.basic_blocks[bb];
+        let result = apply_deaths(bb_data, 0, states[bb]);
+        if result != 0 {
+            for succ in bb_data.terminator().successors() {
+                if succ == location.block {
+                    return None;
+                }
+                if join(&mut states[succ], result) && enqueue(&mut states[succ]) {
+                    queue.push(succ);
+                }
+            }
+        }
+    }
+
+    Some(states)
 }
 
 struct V<'a, const N: usize> {

--- a/clippy_utils/src/mir/mod.rs
+++ b/clippy_utils/src/mir/mod.rs
@@ -1,11 +1,14 @@
 use std::iter;
+use std::ops::Range;
 
 use rustc_data_structures::either::Either;
+use rustc_data_structures::fx::FxHashSet;
 use rustc_hir::{Expr, HirId};
 use rustc_index::bit_set::DenseBitSet;
 use rustc_middle::mir::visit::{MutatingUseContext, NonMutatingUseContext, PlaceContext, Visitor};
 use rustc_middle::mir::{
-    BasicBlock, Body, InlineAsmOperand, Local, Location, Place, START_BLOCK, StatementKind, TerminatorKind, traversal,
+    BasicBlock, Body, InlineAsmOperand, Local, Location, Place, START_BLOCK, Statement, StatementKind, TerminatorKind,
+    traversal,
 };
 use rustc_middle::ty::TyCtxt;
 
@@ -29,33 +32,104 @@ pub fn visit_local_usage<const N: usize>(
     mir: &Body<'_>,
     location: Location,
 ) -> Option<[LocalUsage; N]> {
-    let init = [const {
-        LocalUsage {
-            local_use_locs: Vec::new(),
-            local_consume_or_mutate_locs: Vec::new(),
-        }
-    }; N];
+    let blocks = reachable_while_storage_live(&locals, mir, location)?;
 
-    traversal::Postorder::new(&mir.basic_blocks, location.block, None)
+    let mut v = V {
+        locals: &locals,
+        location,
+        results: [const {
+            LocalUsage {
+                local_use_locs: Vec::new(),
+                local_consume_or_mutate_locs: Vec::new(),
+            }
+        }; N],
+    };
+
+    for tbb in traversal::Postorder::new(&mir.basic_blocks, location.block, None)
         .collect::<Vec<_>>()
         .into_iter()
         .rev()
-        .try_fold(init, |usage, tbb| {
-            let tdata = &mir.basic_blocks[tbb];
+    {
+        if blocks.contains(tbb) {
+            v.visit_basic_block_data(tbb, &mir.basic_blocks[tbb]);
+        }
+    }
 
-            // Give up on loops
-            if tdata.terminator().successors().any(|s| s == location.block) {
-                return None;
-            }
+    Some(v.results)
+}
 
-            let mut v = V {
-                locals: &locals,
-                location,
-                results: usage,
-            };
-            v.visit_basic_block_data(tbb, tdata);
-            Some(v.results)
-        })
+/// The tracked values that are still live, as a set of idx of `locals`
+type Live = DenseBitSet<usize>;
+
+/// The [`Local`] whose value `stmt` kills, if any.
+fn killed_local(stmt: &Statement<'_>) -> Option<Local> {
+    match stmt.kind {
+        StatementKind::StorageDead(local) => Some(local),
+        _ => None,
+    }
+}
+
+/// Returns the blocks reachable from `location` in which `locals` still hold their values.
+///
+/// A path is walked until every local has been `StorageDead`-ed.
+/// No statement past that point can refer to the values being tracked.
+/// A local declared inside a loop is therefore distinct on each iteration.
+///
+/// Returns `None` if `location.block` is reachable again while some local is still live.
+/// The statements preceding `location` then run a second time and cannot be ordered against it.
+fn reachable_while_storage_live<const N: usize>(
+    locals: &[Local; N],
+    mir: &Body<'_>,
+    location: Location,
+) -> Option<DenseBitSet<BasicBlock>> {
+    let remove_killed = |bb: BasicBlock, range: Range<usize>, live: &mut Live| {
+        for slot in mir.basic_blocks[bb].statements[range]
+            .iter()
+            .filter_map(killed_local)
+            .filter_map(|killed| locals.iter().position(|&local| local == killed))
+        {
+            live.remove(slot);
+        }
+    };
+    let successors = |bb: BasicBlock, live: &Live| {
+        mir.basic_blocks[bb]
+            .terminator()
+            .successors()
+            .map(|succ| (succ, live.clone()))
+            .collect::<Vec<_>>()
+    };
+
+    let mut blocks = DenseBitSet::new_empty(mir.basic_blocks.len());
+    blocks.insert(location.block);
+
+    let mut seen = FxHashSet::default();
+    let mut worklist = Vec::new();
+    // `location.block` is the only block entered part-way through, so it is walked separately.
+    // Reaching it again in the loop below means a cycle closed around `location`.
+    let stmts = mir.basic_blocks[location.block].statements.len();
+    let after_location = (location.statement_index + 1).min(stmts);
+    let mut live = Live::new_filled(N);
+    remove_killed(location.block, after_location..stmts, &mut live);
+    if !live.is_empty() {
+        worklist.extend(successors(location.block, &live));
+    }
+
+    while let Some((bb, mut live)) = worklist.pop() {
+        if bb == location.block {
+            return None;
+        }
+        if !seen.insert((bb, live.clone())) {
+            continue;
+        }
+        blocks.insert(bb);
+
+        remove_killed(bb, 0..mir.basic_blocks[bb].statements.len(), &mut live);
+        if !live.is_empty() {
+            worklist.extend(successors(bb, &live));
+        }
+    }
+
+    Some(blocks)
 }
 
 struct V<'a, const N: usize> {

--- a/clippy_utils/src/mir/mod.rs
+++ b/clippy_utils/src/mir/mod.rs
@@ -8,7 +8,6 @@ use rustc_index::bit_set::DenseBitSet;
 use rustc_middle::mir::visit::{MutatingUseContext, NonMutatingUseContext, PlaceContext, Visitor};
 use rustc_middle::mir::{
     BasicBlock, Body, InlineAsmOperand, Local, Location, Place, START_BLOCK, Statement, StatementKind, TerminatorKind,
-    traversal,
 };
 use rustc_middle::ty::TyCtxt;
 
@@ -45,11 +44,7 @@ pub fn visit_local_usage<const N: usize>(
         }; N],
     };
 
-    for tbb in traversal::Postorder::new(&mir.basic_blocks, location.block, None)
-        .collect::<Vec<_>>()
-        .into_iter()
-        .rev()
-    {
+    for &tbb in mir.basic_blocks.reverse_postorder() {
         if blocks.contains(tbb) {
             v.visit_basic_block_data(tbb, &mir.basic_blocks[tbb]);
         }

--- a/tests/ui/redundant_clone.fixed
+++ b/tests/ui/redundant_clone.fixed
@@ -291,3 +291,41 @@ mod issue13900 {
         }
     }
 }
+
+mod issue16096 {
+    // `algo` is declared inside the loop, so it is a fresh value on each iteration and the
+    // clone is redundant regardless of how the iteration ends.
+    fn try_from_bytes(bytes: &[u8]) -> Option<Vec<Vec<u8>>> {
+        let mut pipeline = Vec::new();
+        let mut index = 0;
+        while index < bytes.len() {
+            match bytes[index] {
+                0 => {
+                    let algo = vec![];
+                    pipeline.push(algo);
+                    //~^ redundant_clone
+                },
+                1 => {
+                    let algo = vec![];
+                    pipeline.push(algo);
+                    //~^ redundant_clone
+                    return Some(pipeline);
+                },
+                _ => {},
+            }
+            index += 1;
+        }
+
+        None
+    }
+
+    fn loop_body(n: usize) -> Vec<String> {
+        let mut out = Vec::new();
+        for _ in 0..n {
+            let s = String::new();
+            out.push(s);
+            //~^ redundant_clone
+        }
+        out
+    }
+}

--- a/tests/ui/redundant_clone.rs
+++ b/tests/ui/redundant_clone.rs
@@ -291,3 +291,41 @@ mod issue13900 {
         }
     }
 }
+
+mod issue16096 {
+    // `algo` is declared inside the loop, so it is a fresh value on each iteration and the
+    // clone is redundant regardless of how the iteration ends.
+    fn try_from_bytes(bytes: &[u8]) -> Option<Vec<Vec<u8>>> {
+        let mut pipeline = Vec::new();
+        let mut index = 0;
+        while index < bytes.len() {
+            match bytes[index] {
+                0 => {
+                    let algo = vec![];
+                    pipeline.push(algo.clone());
+                    //~^ redundant_clone
+                },
+                1 => {
+                    let algo = vec![];
+                    pipeline.push(algo.clone());
+                    //~^ redundant_clone
+                    return Some(pipeline);
+                },
+                _ => {},
+            }
+            index += 1;
+        }
+
+        None
+    }
+
+    fn loop_body(n: usize) -> Vec<String> {
+        let mut out = Vec::new();
+        for _ in 0..n {
+            let s = String::new();
+            out.push(s.clone());
+            //~^ redundant_clone
+        }
+        out
+    }
+}

--- a/tests/ui/redundant_clone.stderr
+++ b/tests/ui/redundant_clone.stderr
@@ -180,5 +180,41 @@ note: this value is dropped without further use
 LL |     foo(&x.clone(), move || {
    |          ^
 
-error: aborting due to 15 previous errors
+error: redundant clone
+  --> tests/ui/redundant_clone.rs:305:39
+   |
+LL |                     pipeline.push(algo.clone());
+   |                                       ^^^^^^^^ help: remove this
+   |
+note: this value is dropped without further use
+  --> tests/ui/redundant_clone.rs:305:35
+   |
+LL |                     pipeline.push(algo.clone());
+   |                                   ^^^^
+
+error: redundant clone
+  --> tests/ui/redundant_clone.rs:310:39
+   |
+LL |                     pipeline.push(algo.clone());
+   |                                       ^^^^^^^^ help: remove this
+   |
+note: this value is dropped without further use
+  --> tests/ui/redundant_clone.rs:310:35
+   |
+LL |                     pipeline.push(algo.clone());
+   |                                   ^^^^
+
+error: redundant clone
+  --> tests/ui/redundant_clone.rs:326:23
+   |
+LL |             out.push(s.clone());
+   |                       ^^^^^^^^ help: remove this
+   |
+note: this value is dropped without further use
+  --> tests/ui/redundant_clone.rs:326:22
+   |
+LL |             out.push(s.clone());
+   |                      ^
+
+error: aborting due to 18 previous errors
 

